Panic in Kubernetes backend

```
2017-01-26 01:04:12.636 [PANIC][5429] syncer.go 348: failed to parse annotation: %sunexpected end of JSON input
panic: (*logrus.Entry) (0x16396c0,0xc4204c5e50)

goroutine 23 [running]:
panic(0x16396c0, 0xc4204c5e50)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/projectcalico/felix/go/vendor/github.com/Sirupsen/logrus.Entry.log(0xc42000e140, 0xc42048bd70, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc420276500, ...)
        /go/src/github.com/projectcalico/felix/go/vendor/github.com/Sirupsen/logrus/entry.go:124 +0x423
github.com/projectcalico/felix/go/vendor/github.com/Sirupsen/logrus.(*Entry).Panic(0xc420450730, 0xc4202765d0, 0x1, 0x1)
        /go/src/github.com/projectcalico/felix/go/vendor/github.com/Sirupsen/logrus/entry.go:169 +0x124
github.com/projectcalico/felix/go/vendor/github.com/Sirupsen/logrus.(*Entry).Panicf(0xc420450730, 0x166d07e, 0x2, 0xc4202768a8, 0x1, 0x1)
        /go/src/github.com/projectcalico/felix/go/vendor/github.com/Sirupsen/logrus/entry.go:217 +0x101
github.com/projectcalico/felix/go/vendor/github.com/Sirupsen/logrus.(*Logger).Panicf(0xc42000e140, 0x166d07e, 0x2, 0xc4202768a8, 0x1, 0x1)
        /go/src/github.com/projectcalico/felix/go/vendor/github.com/Sirupsen/logrus/logger.go:172 +0x86
github.com/projectcalico/felix/go/vendor/github.com/Sirupsen/logrus.Panicf(0x166d07e, 0x2, 0xc4202768a8, 0x1, 0x1)
        /go/src/github.com/projectcalico/felix/go/vendor/github.com/Sirupsen/logrus/exported.go:147 +0x5f
github.com/projectcalico/felix/go/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s.(*kubeSyncer).performSnapshot(0xc4204c4230, 0x1, 0xc4205257b8, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/projectcalico/felix/go/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/syncer.go:348 +0xc6f
github.com/projectcalico/felix/go/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s.(*kubeSyncer).readFromKubernetesAPI(0xc4204c4230)
        /go/src/github.com/projectcalico/felix/go/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/syncer.go:149 +0x199c
created by github.com/projectcalico/felix/go/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s.(*kubeSyncer).Start
        /go/src/github.com/projectcalico/felix/go/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/syncer.go:66 +0x3f
```

Seen when setting the Namespace annotations to an invalid value with Calico v2.0.1.  Felix then repeatedly crashes.

My Namespace:

```
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    net.beta.kubernetes.io/network-policy: ""
  creationTimestamp: 2017-01-26T00:30:12Z
  name: cali-ns-0
  resourceVersion: "191151"
  selfLink: /api/v1/namespacescali-ns-0
  uid: 999d96a1-e35e-11e6-85e4-42010a8a0138
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```

Fixing the malformed annotation fixes the crash loop.

--- a/glide.lock
+++ b/glide.lock
@@ -1,5 +1,5 @@
 hash: 283f982821510d6c8c5e48054919974b2b2b9664b838ec9f9a06705e19ed3d99
-updated: 2017-01-19T19:48:14.880969403Z
+updated: 2017-01-26T21:23:50.030310202-08:00
 imports:
 - name: cloud.google.com/go
   version: 3b1ae45394a234c385be014e9a488f2bb6eef821
@@ -9,7 +9,7 @@ imports:
 - name: github.com/blang/semver
   version: 31b736133b98f26d5e078ec9eb591666edfd091f
 - name: github.com/coreos/etcd
-  version: 531c3061c1ffe9e882cd7e0eeb4edb6230f454d0
+  version: d2716fc5ae7909a3b9651e31cfb0eba22b4920c3
   subpackages:
   - client
   - pkg/fileutil
@@ -17,6 +17,7 @@ imports:
   - pkg/tlsutil
   - pkg/transport
   - pkg/types
+  - version
 - name: github.com/coreos/go-oidc
   version: 5644a2f50e2d2d5ba0b474bc5bc55fea1925936d
   subpackages:
@@ -25,6 +26,10 @@ imports:
   - key
   - oauth2
   - oidc
+- name: github.com/coreos/go-semver
+  version: 568e959cd89871e61434c1143528d9162da89ef2
+  subpackages:
+  - semver
 - name: github.com/coreos/go-systemd
   version: 48702e0da86bd25e76cfef347e2adeb434a0d0a6
   subpackages:
@@ -85,7 +90,7 @@ imports:
 - name: github.com/juju/ratelimit
   version: 77ed1c8a01217656d2080ad51981f6e99adaa177
 - name: github.com/kelseyhightower/envconfig
-  version: 5c008110b20b657eb7e005b83d0a5f6aa6bb5f4b
+  version: f611eb38b3875cc3bd991ca91c51d06446afa14c
 - name: github.com/mailru/easyjson
   version: d5b7844b561a7bc640052f1b935f7b800330d7e0
   subpackages:
@@ -93,7 +98,7 @@ imports:
   - jlexer
   - jwriter
 - name: github.com/onsi/ginkgo
-  version: 00054c0bb96fc880d4e0be1b90937fad438c5290
+  version: bb93381d543b0e5725244abe752214a110791d01
   subpackages:
   - config
   - extensions/table
@@ -133,7 +138,7 @@ imports:
 - name: github.com/spf13/pflag
   version: 08b1a584251b5b62f458943640fc8ebd4d50aaa5
 - name: github.com/ugorji/go
-  version: 9c7f9b7a2bc3a520f7c7b30b34b7f85f47fe27b6
+  version: ded73eae5db7e7a0ef6f55aace87a2873c5d2b74
   subpackages:
   - codec
 - name: golang.org/x/crypto
@@ -143,13 +148,14 @@ imports:
   - blowfish
   - ssh/terminal
 - name: golang.org/x/net
-  version: 6acef71eb69611914f7a30939ea9f6e194c78172
+  version: f2499483f923065a842d38eb4c7f1927e6fc6e6d
   subpackages:
   - context
   - context/ctxhttp
   - http2
   - http2/hpack
   - idna
+  - lex/httplex
 - name: golang.org/x/oauth2
   version: 3c3a985cb79f52a3190fbc056984415ca6763d01
   subpackages:
@@ -158,7 +164,7 @@ imports:
   - jws
   - jwt
 - name: golang.org/x/sys
-  version: 8f0908ab3b2457e2e15403d3697c9ef5cb4b57a9
+  version: d75a52659825e75fff6158388dddc6a5b04f9ba5
   subpackages:
   - unix
 - name: golang.org/x/text
@@ -315,7 +321,7 @@ imports:
   - transport
 testImports:
 - name: github.com/onsi/gomega
-  version: f1f0f388b31eca4e2cbe7a6dd8a3a1dfddda5b1c
+  version: e85f63af4302dc0e4277c6008ecf803f1d80e4f0
   subpackages:
   - format
   - internal/assertion

--- a/lib/backend/k8s/conversion.go
+++ b/lib/backend/k8s/conversion.go
@@ -63,13 +63,32 @@ func (c converter) parseWorkloadID(workloadID string) (string, string) {
 	return splits[0], splits[1]
 }
 
-// parsePolicyName extracts the Namespace and NetworkPolicy name from the given Policy name.
-func (c converter) parsePolicyName(name string) (string, string) {
-	splits := strings.SplitN(name, ".", 2)
-	if len(splits) != 2 {
-		return "", ""
+// parsePolicyNameNamespace extracts the Kubernetes Namespace that backs the given Policy.
+func (c converter) parsePolicyNameNamespace(name string) (string, error) {
+	// Policy objects backed by Namespaces have form "ns.projectcalico.org/<ns_name>"
+	if !strings.HasPrefix(name, "ns.projectcalico.org/") {
+		// This is not backed by a Kubernetes Namespace.
+		return "", fmt.Errorf("Policy %s not backed by a Namespace", name)
 	}
-	return splits[0], splits[1]
+
+	return strings.TrimPrefix(name, "ns.projectcalico.org/"), nil
+
+}
+
+// parsePolicyNameNetworkPolicy extracts the Kubernetes Namespace and NetworkPolicy that backs the given Policy.
+func (c converter) parsePolicyNameNetworkPolicy(name string) (string, string, error) {
+	// Policies backed by NetworkPolicies have form "np.projectcalico.org/<ns_name>.<np_name>
+	if !strings.HasPrefix(name, "np.projectcalico.org/") {
+		// This is not backed by a Kubernetes NetworkPolicy.
+		return "", "", fmt.Errorf("Policy %s not backed by a NetworkPolicy", name)
+	}
+
+	splits := strings.SplitN(strings.TrimPrefix(name, "np.projectcalico.org/"), ".", 2)
+	if len(splits) != 2 {
+		return "", "", fmt.Errorf("Name does not include both Namespace and NetworkPolicy: %s", name)
+	}
+	// Return Namespace, NetworkPolicy name.
+	return splits[0], splits[1], nil
 }
 
 // parseProfileName extracts the Namespace name from the given Profile name.
@@ -81,14 +100,19 @@ func (c converter) parseProfileName(profileName string) (string, error) {
 	return splits[1], nil
 }
 
-func (c converter) namespaceToProfile(ns *kapiv1.Namespace) (*model.KVPair, error) {
+// namespaceToPolicy converts a Namespace to a Policy.  We create a Policy per-Namespace
+// to implement per-Namespace ingress behavior (e.g DefaultDeny).  It also ensures that
+// every k8s Pod is selected by at least one Policy that allows egress traffic.
+func (c converter) namespaceToPolicy(ns *kapiv1.Namespace) (*model.KVPair, error) {
 	// Determine the ingress action based off the DefaultDeny annotation.
 	ingressAction := "allow"
 	for k, v := range ns.ObjectMeta.Annotations {
 		if k == policyAnnotation {
 			np := namespacePolicy{}
 			if err := json.Unmarshal([]byte(v), &np); err != nil {
-				return nil, goerrors.New(fmt.Sprint("failed to parse annotation: %s", err))
+				// We want to handle this case gracefully since this can
+				// occur due to user error.
+				log.Warnf("Failed to parse annotation on Namespace '%s'.", ns.Name)
 			}
 			if np.Ingress.Isolation == "DefaultDeny" {
 				ingressAction = "deny"
@@ -96,22 +120,40 @@ func (c converter) namespaceToProfile(ns *kapiv1.Namespace) (*model.KVPair, erro
 		}
 	}
 
-	// Generate the labels to apply to the profile.
+	name := fmt.Sprintf("ns.projectcalico.org/%s", ns.ObjectMeta.Name)
+	kvp := model.KVPair{
+		Key: model.PolicyKey{Name: name},
+		Value: &model.Policy{
+			Selector:      fmt.Sprintf("calico/k8s_ns == '%s'", ns.Name),
+			InboundRules:  []model.Rule{model.Rule{Action: ingressAction}},
+			OutboundRules: []model.Rule{model.Rule{Action: "allow"}},
+		},
+		Revision: ns.ObjectMeta.ResourceVersion,
+	}
+	return &kvp, nil
+}
+
+// namespaceToProfile converts a Namespace to a Calico Profile.  The Profile stores
+// labels from the Namespace which are inherited by the WorkloadEndpoints within
+// the Profile, however no rules are populated.  Per-Namespace network
+// policy rules are implemented in namespaceToPolicy.
+func (c converter) namespaceToProfile(ns *kapiv1.Namespace) (*model.KVPair, error) {
+	// Generate the labels to apply to the profile, using a special prefix
+	// to indicate that these are the labels from the parent Kubernetes Namespace.
 	labels := map[string]string{}
 	for k, v := range ns.ObjectMeta.Labels {
 		labels[fmt.Sprintf("k8s_ns/label/%s", k)] = v
 	}
 
-	name := fmt.Sprintf("k8s_ns.%s", ns.ObjectMeta.Name)
+	name := fmt.Sprintf("ns.projectcalico.org/%s", ns.ObjectMeta.Name)
 	kvp := model.KVPair{
 		Key: model.ProfileKey{Name: name},
 		Value: &model.Profile{
-			Rules: model.ProfileRules{
-				InboundRules:  []model.Rule{model.Rule{Action: ingressAction}},
-				OutboundRules: []model.Rule{model.Rule{Action: "allow"}},
-			},
-			Tags:   []string{name},
 			Labels: labels,
+			Rules: model.ProfileRules{
+				InboundRules:  []model.Rule{},
+				OutboundRules: []model.Rule{},
+			},
 		},
 		Revision: ns.ObjectMeta.ResourceVersion,
 	}
@@ -162,7 +204,7 @@ func (c converter) hasIPAddress(pod *kapiv1.Pod) bool {
 
 func (c converter) podToWorkloadEndpoint(pod *kapiv1.Pod) (*model.KVPair, error) {
 	// Pull out the profile and workload ID based on pod name and Namespace.
-	profile := fmt.Sprintf("k8s_ns.%s", pod.ObjectMeta.Namespace)
+	profile := fmt.Sprintf("ns.projectcalico.org/%s", pod.ObjectMeta.Namespace)
 	workload := fmt.Sprintf("%s.%s", pod.ObjectMeta.Namespace, pod.ObjectMeta.Name)
 
 	// If the pod doesn't have an IP address yet, then it hasn't gone through CNI.
@@ -211,7 +253,7 @@ func (c converter) podToWorkloadEndpoint(pod *kapiv1.Pod) (*model.KVPair, error)
 // networkPolicyToPolicy converts a k8s NetworkPolicy to a model.KVPair.
 func (c converter) networkPolicyToPolicy(np *extensions.NetworkPolicy) (*model.KVPair, error) {
 	// Pull out important fields.
-	policyName := fmt.Sprintf("%s.%s", np.ObjectMeta.Namespace, np.ObjectMeta.Name)
+	policyName := fmt.Sprintf("np.projectcalico.org/%s.%s", np.ObjectMeta.Namespace, np.ObjectMeta.Name)
 	order := float64(1000.0)
 
 	// Generate the inbound rules list.

--- a/lib/backend/k8s/conversion_test.go
+++ b/lib/backend/k8s/conversion_test.go
@@ -37,11 +37,35 @@ var _ = Describe("Test parsing strings", func() {
 		Expect(podName).To(Equal("podName"))
 	})
 
-	It("should parse policy names", func() {
-		name := "Namespace.policyName"
-		ns, polName := c.parsePolicyName(name)
+	It("should parse valid policy names", func() {
+		// Parse a NetworkPolicy backed Policy.
+		name := "np.projectcalico.org/Namespace.policyName"
+		ns, polName, err := c.parsePolicyNameNetworkPolicy(name)
+		Expect(err).NotTo(HaveOccurred())
 		Expect(ns).To(Equal("Namespace"))
 		Expect(polName).To(Equal("policyName"))
+
+		// Parse a Namespace backed Policy.
+		name = "ns.projectcalico.org/Namespace"
+		ns, err = c.parsePolicyNameNamespace(name)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(ns).To(Equal("Namespace"))
+
+	})
+
+	It("should not parse invalid policy names", func() {
+		name := "something.projectcalico.org/Namespace.Name"
+
+		// As a NetworkPolicy.
+		ns, polName, err := c.parsePolicyNameNetworkPolicy(name)
+		Expect(err).To(HaveOccurred())
+		Expect(ns).To(Equal(""))
+		Expect(polName).To(Equal(""))
+
+		// As a Namespace.
+		ns, err = c.parsePolicyNameNamespace(name)
+		Expect(err).To(HaveOccurred())
+		Expect(ns).To(Equal(""))
 	})
 
 	It("should parse profile names", func() {
@@ -206,7 +230,7 @@ var _ = Describe("Test NetworkPolicy conversion", func() {
 		Expect(err).NotTo(HaveOccurred())
 
 		// Assert key fields are correct.
-		Expect(pol.Key.(model.PolicyKey).Name).To(Equal("default.testPolicy"))
+		Expect(pol.Key.(model.PolicyKey).Name).To(Equal("np.projectcalico.org/default.testPolicy"))
 
 		// Assert value fields are correct.
 		Expect(int(*pol.Value.(*model.Policy).Order)).To(Equal(1000))
@@ -234,7 +258,7 @@ var _ = Describe("Test NetworkPolicy conversion", func() {
 		Expect(err).NotTo(HaveOccurred())
 
 		// Assert key fields are correct.
-		Expect(pol.Key.(model.PolicyKey).Name).To(Equal("default.testPolicy"))
+		Expect(pol.Key.(model.PolicyKey).Name).To(Equal("np.projectcalico.org/default.testPolicy"))
 
 		// Assert value fields are correct.
 		Expect(int(*pol.Value.(*model.Policy).Order)).To(Equal(1000))
@@ -259,7 +283,7 @@ var _ = Describe("Test NetworkPolicy conversion", func() {
 		Expect(err).NotTo(HaveOccurred())
 
 		// Assert key fields are correct.
-		Expect(pol.Key.(model.PolicyKey).Name).To(Equal("default.testPolicy"))
+		Expect(pol.Key.(model.PolicyKey).Name).To(Equal("np.projectcalico.org/default.testPolicy"))
 
 		// Assert value fields are correct.
 		Expect(int(*pol.Value.(*model.Policy).Order)).To(Equal(1000))
@@ -292,7 +316,7 @@ var _ = Describe("Test NetworkPolicy conversion", func() {
 		Expect(err).NotTo(HaveOccurred())
 
 		// Assert key fields are correct.
-		Expect(pol.Key.(model.PolicyKey).Name).To(Equal("default.testPolicy"))
+		Expect(pol.Key.(model.PolicyKey).Name).To(Equal("np.projectcalico.org/default.testPolicy"))
 
 		// Assert value fields are correct.
 		Expect(int(*pol.Value.(*model.Policy).Order)).To(Equal(1000))
@@ -329,7 +353,7 @@ var _ = Describe("Test NetworkPolicy conversion", func() {
 		Expect(err).NotTo(HaveOccurred())
 
 		// Assert key fields are correct.
-		Expect(pol.Key.(model.PolicyKey).Name).To(Equal("default.testPolicy"))
+		Expect(pol.Key.(model.PolicyKey).Name).To(Equal("np.projectcalico.org/default.testPolicy"))
 
 		// Assert value fields are correct.
 		Expect(int(*pol.Value.(*model.Policy).Order)).To(Equal(1000))
@@ -347,10 +371,14 @@ var _ = Describe("Test Namespace conversion", func() {
 	// Use a single instance of the converter for these tests.
 	c := converter{}
 
-	It("should parse a Namespace to a Profile with allow rules", func() {
+	It("should parse a Namespace to a Profile", func() {
 		ns := k8sapi.Namespace{
 			ObjectMeta: k8sapi.ObjectMeta{
-				Name:        "default",
+				Name: "default",
+				Labels: map[string]string{
+					"foo":   "bar",
+					"roger": "rabbit",
+				},
 				Annotations: map[string]string{},
 			},
 			Spec: k8sapi.NamespaceSpec{},
@@ -359,16 +387,47 @@ var _ = Describe("Test Namespace conversion", func() {
 		p, err := c.namespaceToProfile(&ns)
 		Expect(err).NotTo(HaveOccurred())
 
-		// Ensure rules are correct.
+		// Ensure rules are correct for profile.
 		inboundRules := p.Value.(*model.Profile).Rules.InboundRules
 		outboundRules := p.Value.(*model.Profile).Rules.OutboundRules
-		Expect(len(inboundRules)).To(Equal(1))
-		Expect(len(outboundRules)).To(Equal(1))
-		Expect(inboundRules[0].Action).To(Equal("allow"))
-		Expect(outboundRules[0].Action).To(Equal("allow"))
+		Expect(len(inboundRules)).To(Equal(0))
+		Expect(len(outboundRules)).To(Equal(0))
+
+		// Check labels.
+		labels := p.Value.(*model.Profile).Labels
+		Expect(labels["k8s_ns/label/foo"]).To(Equal("bar"))
+		Expect(labels["k8s_ns/label/roger"]).To(Equal("rabbit"))
 	})
 
-	It("should parse a Namespace to a Profile with deny rules", func() {
+	It("should parse a Namespace to a Policy", func() {
+		ns := k8sapi.Namespace{
+			ObjectMeta: k8sapi.ObjectMeta{
+				Name: "default",
+				Labels: map[string]string{
+					"foo":   "bar",
+					"roger": "rabbit",
+				},
+				Annotations: map[string]string{},
+			},
+			Spec: k8sapi.NamespaceSpec{},
+		}
+
+		// Ensure it generates the correct Policy.
+		kvp, err := c.namespaceToPolicy(&ns)
+		Expect(err).NotTo(HaveOccurred())
+		key := kvp.Key.(model.PolicyKey)
+		policy := kvp.Value.(*model.Policy)
+		Expect(key.Name).To(Equal("ns.projectcalico.org/default"))
+		Expect(policy.Selector).To(Equal("calico/k8s_ns == 'default'"))
+		Expect(len(policy.InboundRules)).To(Equal(1))
+		Expect(len(policy.OutboundRules)).To(Equal(1))
+
+		allow := model.Rule{Action: "allow"}
+		Expect(policy.InboundRules[0]).To(Equal(allow))
+		Expect(policy.OutboundRules[0]).To(Equal(allow))
+	})
+
+	It("should parse a Namespace to a Profile with no labels", func() {
 		ns := k8sapi.Namespace{
 			ObjectMeta: k8sapi.ObjectMeta{
 				Name: "default",
@@ -385,13 +444,43 @@ var _ = Describe("Test Namespace conversion", func() {
 		// Ensure rules are correct.
 		inboundRules := p.Value.(*model.Profile).Rules.InboundRules
 		outboundRules := p.Value.(*model.Profile).Rules.OutboundRules
-		Expect(len(inboundRules)).To(Equal(1))
-		Expect(len(outboundRules)).To(Equal(1))
-		Expect(inboundRules[0].Action).To(Equal("deny"))
-		Expect(outboundRules[0].Action).To(Equal("allow"))
+		Expect(len(inboundRules)).To(Equal(0))
+		Expect(len(outboundRules)).To(Equal(0))
+
+		// Check labels.
+		labels := p.Value.(*model.Profile).Labels
+		Expect(len(labels)).To(Equal(0))
 	})
 
-	It("should fail for invalid annotation", func() {
+	It("should parse a Namespace to Policy (DefaultDeny)", func() {
+		ns := k8sapi.Namespace{
+			ObjectMeta: k8sapi.ObjectMeta{
+				Name: "default",
+				Annotations: map[string]string{
+					"net.beta.kubernetes.io/network-policy": "{\"ingress\": {\"isolation\": \"DefaultDeny\"}}",
+				},
+			},
+			Spec: k8sapi.NamespaceSpec{},
+		}
+
+		// Ensure it generates the correct Policy.
+		kvp, err := c.namespaceToPolicy(&ns)
+		Expect(err).NotTo(HaveOccurred())
+		key := kvp.Key.(model.PolicyKey)
+		policy := kvp.Value.(*model.Policy)
+		Expect(key.Name).To(Equal("ns.projectcalico.org/default"))
+		Expect(policy.Selector).To(Equal("calico/k8s_ns == 'default'"))
+		Expect(len(policy.InboundRules)).To(Equal(1))
+		Expect(len(policy.OutboundRules)).To(Equal(1))
+
+		allow := model.Rule{Action: "allow"}
+		deny := model.Rule{Action: "deny"}
+		Expect(policy.InboundRules[0]).To(Equal(deny))
+		Expect(policy.OutboundRules[0]).To(Equal(allow))
+
+	})
+
+	It("should not fail for malformed annotation", func() {
 		ns := k8sapi.Namespace{
 			ObjectMeta: k8sapi.ObjectMeta{
 				Name: "default",
@@ -402,8 +491,16 @@ var _ = Describe("Test Namespace conversion", func() {
 			Spec: k8sapi.NamespaceSpec{},
 		}
 
-		_, err := c.namespaceToProfile(&ns)
-		Expect(err).To(HaveOccurred())
+		By("converting to a Profile", func() {
+			_, err := c.namespaceToProfile(&ns)
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		By("converting to a Policy", func() {
+			_, err := c.namespaceToPolicy(&ns)
+			Expect(err).NotTo(HaveOccurred())
+		})
+
 	})
 
 	It("should handle a valid but not DefaultDeny annotation", func() {
@@ -417,15 +514,21 @@ var _ = Describe("Test Namespace conversion", func() {
 			Spec: k8sapi.NamespaceSpec{},
 		}
 
-		p, err := c.namespaceToProfile(&ns)
-		Expect(err).NotTo(HaveOccurred())
+		By("converting to a Profile", func() {
+			p, err := c.namespaceToProfile(&ns)
+			Expect(err).NotTo(HaveOccurred())
 
-		// Ensure rules are correct.
-		inboundRules := p.Value.(*model.Profile).Rules.InboundRules
-		outboundRules := p.Value.(*model.Profile).Rules.OutboundRules
-		Expect(len(inboundRules)).To(Equal(1))
-		Expect(len(outboundRules)).To(Equal(1))
-		Expect(inboundRules[0].Action).To(Equal("allow"))
-		Expect(outboundRules[0].Action).To(Equal("allow"))
+			// Ensure rules are correct.
+			inboundRules := p.Value.(*model.Profile).Rules.InboundRules
+			outboundRules := p.Value.(*model.Profile).Rules.OutboundRules
+			Expect(len(inboundRules)).To(Equal(0))
+			Expect(len(outboundRules)).To(Equal(0))
+		})
+
+		By("converting to a Policy", func() {
+			_, err := c.namespaceToPolicy(&ns)
+			Expect(err).NotTo(HaveOccurred())
+		})
+
 	})
 })

--- a/lib/backend/k8s/k8s.go
+++ b/lib/backend/k8s/k8s.go
@@ -543,6 +543,19 @@ func (c *KubeClient) listPolicies(l model.PolicyListOptions) ([]*model.KVPair, e
 		ret = append(ret, kvp)
 	}
 
+	// List all Namespaces and turn them into Policies as well.
+	namespaces, err := c.clientSet.Namespaces().List(kapiv1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	for _, ns := range namespaces.Items {
+		kvp, err := c.converter.namespaceToPolicy(&ns)
+		if err != nil {
+			return nil, err
+		}
+		ret = append(ret, kvp)
+	}
+
 	return ret, nil
 }
 
@@ -551,20 +564,44 @@ func (c *KubeClient) getPolicy(k model.PolicyKey) (*model.KVPair, error) {
 	if k.Name == "" {
 		return nil, goerrors.New("Missing policy name")
 	}
-	namespace, policyName := c.converter.parsePolicyName(k.Name)
 
-	networkPolicy := extensions.NetworkPolicy{}
-	err := c.clientSet.Extensions().RESTClient().
-		Get().
-		Resource("networkpolicies").
-		Namespace(namespace).
-		Name(policyName).
-		Timeout(10 * time.Second).
-		Do().Into(&networkPolicy)
-	if err != nil {
-		return nil, resources.K8sErrorToCalico(err, k)
+	// Check to see if this is backed by a NetworkPolicy or a Namespace.
+	if strings.HasPrefix(k.Name, "np.projectcalico.org/") {
+		// Backed by a NetworkPolicy. Parse out the namespace / name.
+		namespace, policyName, err := c.converter.parsePolicyNameNetworkPolicy(k.Name)
+		if err != nil {
+			return nil, errors.ErrorResourceDoesNotExist{Err: err, Identifier: k}
+		}
+
+		// Get the NetworkPolicy from the API and convert it.
+		networkPolicy := extensions.NetworkPolicy{}
+		err = c.clientSet.Extensions().RESTClient().
+			Get().
+			Resource("networkpolicies").
+			Namespace(namespace).
+			Name(policyName).
+			Timeout(10 * time.Second).
+			Do().Into(&networkPolicy)
+		if err != nil {
+			return nil, resources.K8sErrorToCalico(err, k)
+		}
+		return c.converter.networkPolicyToPolicy(&networkPolicy)
+	} else if strings.HasPrefix(k.Name, "ns.projectcalico.org/") {
+		// This is backed by a Namespace.
+		namespace, err := c.converter.parsePolicyNameNamespace(k.Name)
+		if err != nil {
+			return nil, errors.ErrorResourceDoesNotExist{Err: err, Identifier: k}
+		}
+
+		ns, err := c.clientSet.Namespaces().Get(namespace, metav1.GetOptions{})
+		if err != nil {
+			return nil, resources.K8sErrorToCalico(err, k)
+		}
+		return c.converter.namespaceToPolicy(ns)
+	} else {
+		// Received a Get() for a Policy that doesn't exist.
+		return nil, errors.ErrorResourceDoesNotExist{Identifier: k}
 	}
-	return c.converter.networkPolicyToPolicy(&networkPolicy)
 }
 
 func (c *KubeClient) getReadyStatus(k model.ReadyFlagKey) (*model.KVPair, error) {
@@ -681,7 +718,10 @@ func (c *KubeClient) deleteGlobalConfig(k *model.KVPair) error {
 }
 
 func (c *KubeClient) getHostConfig(k model.HostConfigKey) (*model.KVPair, error) {
-	return nil, goerrors.New("Get for HostConfig not supported in kubernetes backend")
+	return &model.KVPair{
+		Key:   k,
+		Value: nil,
+	}, nil
 }
 
 func (c *KubeClient) listHostConfig(l model.HostConfigListOptions) ([]*model.KVPair, error) {

--- a/lib/backend/k8s/k8s_fv_test.go
+++ b/lib/backend/k8s/k8s_fv_test.go
@@ -44,16 +44,17 @@ func (c cb) OnUpdates(updates []api.Update) {
 	for _, u := range updates {
 		switch u.UpdateType {
 		case api.UpdateTypeKVNew:
-			// Value should not be nil.
-			Expect(u.Value).NotTo(BeNil())
+			// Sometimes the value is nil (e.g ProfileTags)
+			log.Infof("[TEST] Syncer received new: %+v", u)
 		case api.UpdateTypeKVUpdated:
-			// Value should not be nil.
-			Expect(u.Value).NotTo(BeNil())
+			// Sometimes the value is nil (e.g ProfileTags)
+			log.Infof("[TEST] Syncer received updated: %+v", u)
 		case api.UpdateTypeKVDeleted:
-			// Ensure the value is nil.
+			// Ensure the value is nil for deletes.
+			log.Infof("[TEST] Syncer received deleted: %+v", u)
 			Expect(u.Value).To(BeNil())
 		case api.UpdateTypeKVUnknown:
-			panic(fmt.Sprintf("Received unkown update: %+v", u))
+			panic(fmt.Sprintf("[TEST] Syncer received unkown update: %+v", u))
 		}
 	}
 }
@@ -113,9 +114,16 @@ var _ = Describe("Test Syncer API for Kubernetes backend", func() {
 		_, err = c.List(model.ProfileListOptions{})
 		Expect(err).NotTo(HaveOccurred())
 
+		_, err = c.List(model.PolicyListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
 		// Perform a Get and ensure no error in the Calico API.
 		_, err = c.Get(model.ProfileKey{Name: fmt.Sprintf("default.%s", ns.ObjectMeta.Name)})
 		Expect(err).NotTo(HaveOccurred())
+
+		_, err = c.Get(model.PolicyKey{Name: fmt.Sprintf("ns.projectcalico.org/%s", ns.ObjectMeta.Name)})
+		Expect(err).NotTo(HaveOccurred())
+
 	})
 
 	It("should handle a Namespace without DefaultDeny", func() {
@@ -139,12 +147,26 @@ var _ = Describe("Test Syncer API for Kubernetes backend", func() {
 		Expect(err).NotTo(HaveOccurred())
 
 		// Perform a List and ensure it shows up in the Calico API.
-		_, err = c.List(model.ProfileListOptions{})
-		Expect(err).NotTo(HaveOccurred())
+		By("listing Profiles", func() {
+			_, err = c.List(model.ProfileListOptions{})
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		By("listing Policies", func() {
+			_, err = c.List(model.PolicyListOptions{})
+			Expect(err).NotTo(HaveOccurred())
+		})
 
 		// Perform a Get and ensure no error in the Calico API.
-		_, err = c.Get(model.ProfileKey{Name: fmt.Sprintf("default.%s", ns.ObjectMeta.Name)})
-		Expect(err).NotTo(HaveOccurred())
+		By("getting a Profile", func() {
+			_, err = c.Get(model.ProfileKey{Name: fmt.Sprintf("default.%s", ns.ObjectMeta.Name)})
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		By("getting a Policy", func() {
+			_, err = c.Get(model.PolicyKey{Name: fmt.Sprintf("ns.projectcalico.org/%s", ns.ObjectMeta.Name)})
+			Expect(err).NotTo(HaveOccurred())
+		})
 	})
 
 	It("should handle a basic NetworkPolicy", func() {
@@ -200,7 +222,7 @@ var _ = Describe("Test Syncer API for Kubernetes backend", func() {
 		Expect(err).NotTo(HaveOccurred())
 
 		// Perform a Get and ensure no error in the Calico API.
-		_, err = c.Get(model.PolicyKey{Name: fmt.Sprintf("default.%s", np.ObjectMeta.Name)})
+		_, err = c.Get(model.PolicyKey{Name: fmt.Sprintf("np.projectcalico.org/default.%s", np.ObjectMeta.Name)})
 		Expect(err).NotTo(HaveOccurred())
 	})
 

--- a/lib/backend/k8s/syncer.go
+++ b/lib/backend/k8s/syncer.go
@@ -352,10 +352,17 @@ func (syn *kubeSyncer) performSnapshot() ([]model.KVPair, map[string]bool, resou
 			tags.Revision = profile.Revision
 			labels.Revision = profile.Revision
 
-			snap = append(snap, *rules, *tags, *labels)
+			// Also create a Policy for this Namespace.
+			policy, err := syn.kc.converter.namespaceToPolicy(&ns)
+			if err != nil {
+				log.Panicf("%s", err)
+			}
+
+			snap = append(snap, *rules, *tags, *labels, *policy)
 			keys[rules.Key.String()] = true
 			keys[tags.Key.String()] = true
 			keys[labels.Key.String()] = true
+			keys[policy.Key.String()] = true
 		}
 
 		// Get NetworkPolicies (Policies)
@@ -467,15 +474,22 @@ func (syn *kubeSyncer) parseNamespaceEvent(e watch.Event) []model.KVPair {
 	tags.Revision = profile.Revision
 	labels.Revision = profile.Revision
 
+	// Convert the Namespace into a policy KVPair.
+	policy, err := syn.kc.converter.namespaceToPolicy(ns)
+	if err != nil {
+		log.Panicf("%s", err)
+	}
+
 	// For deletes, we need to nil out the Value part of the KVPair.
 	if e.Type == watch.Deleted {
 		rules.Value = nil
 		tags.Value = nil
 		labels.Value = nil
+		policy.Value = nil
 	}
 
 	// Return the updates.
-	return []model.KVPair{*rules, *tags, *labels}
+	return []model.KVPair{*rules, *tags, *labels, *policy}
 }
 
 // parsePodEvent returns a KVPair for the given event.  If the event isn't

